### DISCUSSION TITLE
MCP23017 i2c 4.7k or 10k resistor?

### DISCUSSION THREAD

**Discussion by JuanFCulliganpy** [tttapa/Control-Surface#1067] [General]
Hi, firstable, i will congra this team or guy for this awesome job, its a really nice librery, im a bit new on that, and i will know how to conect the mcp23017 (i2c via) to arduino pro micro, i know sda and clk wires into arduino respectly, but, on couple of forums said that i need conect resistor between both wires 4.7 o 10k? its a really noob question and sorry for that, another question its, if i need 2 mcp23017 (i wanna conect 16 encoders) and i saw this code 
"MCPEncoderType enc {Wire, 0x0, 12};
//                  │     │    └─ Interrupt pin
//                  │     └────── Address offset
//                  └──────────── I²C interface
 
// Instantiate 8 MIDI rotary encoders.
CCMCPEncoder ccencoders[] {
  {
    enc[0],       // The encoder to use
    [MCU::V_POT_1](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#ga9400848eec9735623e1502aef5bb7cb3), // The MIDI address
    1,            // Encoder speed multiplier
    4,            // Number of pulses per physical "click" of the encoder
  },
  {enc[1], [MCU::V_POT_2](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#ga92a2c8cd9cbff4a39644e3d056a8d4f2)},
  {enc[2], [MCU::V_POT_3](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#gae2f0151f08f23de10baefa5569fe5f43)},
  {enc[3], [MCU::V_POT_4](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#gaf8267cd2589cdf181931744188583a25)},
  {enc[4], [MCU::V_POT_5](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#ga35d2436eb27c2895c585b665167a0f1b)},
  {enc[5], [MCU::V_POT_6](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#gac6ecffbdbfd1a0f59d3894932806e354)},
  {enc[6], [MCU::V_POT_7](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#ga965a18e233916727bb793a7c777603f7)},
  {enc[7], [MCU::V_POT_8](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d7e/group__MCU__CC.html#gaa315e61ff0ceca932be00db242990090)},
};
 
void setup() {
  [Control_Surface](https://tttapa.github.io/Control-Surface-doc/Doxygen/d9/d38/Control__Surface__Class_8cpp.html#af7a5d250297d82b2b29605dd7ac250df).[begin](https://tttapa.github.io/Control-Surface-doc/Doxygen/dc/d45/classControl__Surface__.html#ab0bdf5cca484fb2ba637c39384b27fb2)();
  Wire.[begin](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d32/structAH_1_1Array.html#a90a285d3deb303b8ce684a1e4fdfe032)(); // Must be called before enc.begin()
  Wire.setClock(800000);
  enc.[begin](https://tttapa.github.io/Control-Surface-doc/Doxygen/d0/d32/structAH_1_1Array.html#a90a285d3deb303b8ce684a1e4fdfe032)(); // Initialize the MCP23017
}" i need create 2 instance of this? sorry i am new, and thanks for the pacience

**Comment by tttapa** [comment]
I'm afraid a discussion of I²C pull-up resistor strength is beyond the scope of this project, but you can find guidelines online, e.g. https://www.ti.com/lit/an/slva689/slva689.pdf.

If you want 16 encoders, you'll have to create multiple instances of the `MCPEncoderType` type. You can create an array of two elements, for example:

```cpp
MCPEncoderType encs[2]{
    {Wire, 0x0, 12}, // Make sure they have different addresses
    {Wire, 0x1, 12},
};

CCMCPEncoder ccencoders[] {
  {encs[0][0], {MCU::V_POT_1, Channel_1}},
  {encs[0][1], {MCU::V_POT_2, Channel_1}},
  {encs[0][2], {MCU::V_POT_3, Channel_1}},
  {encs[0][3], {MCU::V_POT_4, Channel_1}},
  {encs[0][4], {MCU::V_POT_5, Channel_1}},
  {encs[0][5], {MCU::V_POT_6, Channel_1}},
  {encs[0][6], {MCU::V_POT_7, Channel_1}},
  {encs[0][7], {MCU::V_POT_8, Channel_1}},
  {encs[1][0], {MCU::V_POT_1, Channel_2}},
  {encs[1][1], {MCU::V_POT_2, Channel_2}},
  {encs[1][2], {MCU::V_POT_3, Channel_2}},
  {encs[1][3], {MCU::V_POT_4, Channel_2}},
  {encs[1][4], {MCU::V_POT_5, Channel_2}},
  {encs[1][5], {MCU::V_POT_6, Channel_2}},
  {encs[1][6], {MCU::V_POT_7, Channel_2}},
  {encs[1][7], {MCU::V_POT_8, Channel_2}},
};


void setup() {
  Control_Surface.begin();
  Wire.begin(); // Must be called before enc.begin()
  Wire.setClock(800000);
  for (auto &enc : encs)
    enc.begin(); // Initialize the MCP23017
}

void loop() {
  for (auto &enc : encs)
    enc.update();
  Control_Surface.loop();
}
```

**Comment by JuanFCulliganpy** [comment]
hi, sorry for the hypernoob question that im going to do, but, i need some help, i conect the mcp23017 to an arduino pro micro i2c via (2 scl, 3 sda) its really necesary put the resistors into both wires? can someone helpme with a simple conecction esquematic? im really lost XD.
thanks in advice.
Take care.
Juan.

**Reply by JuanFCulliganpy** [reply]
here is my squematic 
![image](https://github.com/user-attachments/assets/f177ab2d-61eb-4c8d-90a0-1f0469f9307d)

**Reply by tttapa** [reply]
Yes, pull-up resistors on the I²C bus (both SDA and SCL) are really necessary.  
I'm afraid that I cannot offer any hardware troubleshooting support at the moment. I'd recommend opening a thread on the Arduino forum (https://forum.arduino.cc) if you need further help.
